usaspending_api/etl/management/commands/check_es_doc_duplication.py

```diff
@@ -0,0 +1,245 @@
+import logging
+import numpy as np
+
+from django.core.management.base import BaseCommand
+from elasticsearch import Elasticsearch
+from elasticsearch.helpers import scan
+from math import ceil, floor
+from multiprocessing.pool import ThreadPool
+from threading import Lock, get_ident
+
+
+_log = logging.getLogger(__name__)
+
+
+class Command(BaseCommand):
+    """Check whether an index has duplicate documents on different shards with the same ``_id`` value
+
+    Assumptions:
+        - The duplication is determined by two docs existing with the same value for meta field ``_id``
+        - There is some non-default shard-routing on the index that would allow for two docs with the same ``_id``
+          value to end up on different shards (by default, shard-routing is done using the ``_id`` field)
+    """
+
+    help = """
+        Check whether an index has duplicate documents on different shards with the same ``_id`` value.
+        There are two approaches, and one needs to be chosen using either --query-type scroll, or --query-type 
+        partition. See arg info for --query-type for more info. There is no fast way to perform this check, 
+        but these two strategies provide a client-side-intense way to do it or a server-side intense way to do it, 
+        respectively. Each takes a very long time to process (e.g. 8hrs for 100M docs). Trial out performance and 
+        impact to client or server by testing with a small --stop-after value.
+    """
+
+    _es_client_config: dict = None
+    _index: str = None
+    _partition_size: int = None
+    _num_partitions: int = None
+    _lock = Lock()
+    _completed_count: int = 0
+    _percents = {p for p in range(1, 101)}
+    _duplicated_doc_ids: dict = {}
+    _duplicated_query_template = {
```

Comment:
Is it necessary to name everything with a trailing underscore? This is a management command, the likelihood of it being imported by other code is incredibly small and discouraged.

Reply:
It's a good practice we should follow everywhere in our python code. If fields are intended to only be used within a class itself (aka "private" in an object-oriented sense), they should be marked via a naming convention as such.

Also generates IDE warnings (as if the underscore wasn't enough) when you are using a "private" field.